presto-main/src/main/java/com/facebook/presto/server/remotetask/WorkerTaskStatusFetcher.java

```diff
@@ -0,0 +1,283 @@
+
+        this.executor = requireNonNull(executor, "executor is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+
+        this.errorTracker = new RequestErrorTracker(taskId, initialTaskStatus.getSelf(), maxErrorDuration, errorScheduledExecutor, "getting task status");
+        this.stats = requireNonNull(stats, "stats is null");
+        this.isBinaryTransportEnabled = isBinaryTransportEnabled;
+    }
+
+    public synchronized void start()
+    {
+        if (running) {
+            // already running
+            return;
+        }
+        running = true;
+        scheduleNextRequest();
+    }
+
+    public synchronized void stop()
+    {
+        running = false;
+        if (future != null) {
+            // do not terminate if the request is already running to avoid closing pooled connections
+            future.cancel(false);
+            future = null;
+        }
+    }
+
+    private synchronized void scheduleNextRequest()
+    {
+        // stopped or done?
+        TaskStatus taskStatus = getTaskStatus();
+        if (!running || taskStatus.getState().isDone()) {
+            return;
+        }
+
+        // outstanding request?
+        if (future != null && !future.isDone()) {
+            // this should never happen
+            log.error("Can not reschedule update because an update is already running");
+            return;
+        }
+
+        // if throttled due to error, asynchronously wait for timeout and try again
+        ListenableFuture<?> errorRateLimit = errorTracker.acquireRequestPermit();
+        if (!errorRateLimit.isDone()) {
+            errorRateLimit.addListener(this::scheduleNextRequest, executor);
+            return;
+        }
+
+        Request request = setContentTypeHeaders(isBinaryTransportEnabled, prepareGet())
+                .setUri(uriBuilderFrom(taskStatus.getSelf()).appendPath("status").build())
+                .setHeader(PRESTO_CURRENT_STATE, taskStatus.getState().toString())
+                .setHeader(PRESTO_MAX_WAIT, refreshMaxWait.toString())
+                .build();
+
+        ResponseHandler responseHandler;
+        if (isBinaryTransportEnabled) {
+            ArrayList<SmileCodec<TaskStatus>> toPass = new ArrayList<>();
+            toPass.add((SmileCodec<TaskStatus>) taskStatusCodec);
+
+            responseHandler = createFullSmileResponseHandler((SmileCodec<List<TaskStatus>>) toPass);
+        }
+        else {
+            responseHandler = createAdaptingJsonResponseHandler(unwrapJsonCodec(taskStatusCodec));
+        }
+
+        errorTracker.startRequest();
+        future = httpClient.executeAsync(request, responseHandler);
+        currentRequestStartNanos.set(System.nanoTime());
+        Futures.addCallback(future, new SimpleHttpResponseHandler<>(this, request.getUri(), stats), executor);
+    }
+
+    TaskStatus getTaskStatus()
+    {
+        return taskStatus.get();
+    }
+
+    @Override
+    public void success(TaskStatus value)
```

Comment:
let's decide how we expect the worker "/v1/task/status?session id" endpoint going to return and handle the response accordingly.
i.e. response = Map of taskId : Taskstatus.